### PR TITLE
APERTA-13009 Export cards as HTML; Include missing context

### PR DESCRIPTION
**Pull request by noxmwalsh**
Signed-off-by: Michael Walsh <noxiousthing@gmail.com>

JIRA issue: https://jira.plos.org/jira/browse/APERTA-13009

#### What this PR does:

Updates the exports.rake task to export a lot of cards with missing information (primarily Author information) and converting the CSV generation into HTML.  This is to allow consumers of these documents to better read and understand the primary gist of a card.

The HTML generation has really basic styling that is mainly used to separate unique sets of data together (for example, an author from the author card will have all of their information in one "box"). 

#### Special instructions for Review or PO:

I highly recommend you get a diverse set of papers to look at in various states (which does exist to an extend with existing seeds).  Even better, if there is a dump from a real-life production use-case, that would be even better.

After a dump is retrieved, run the following Rake task for a semi-random sampling of papers in different publishing states:

```sh
bundle exec rake export:random_manuscript_zips
```

For just a single specific paper, just pass in doi into this Rake task:

```sh
bundle exec rake export:manuscript_zip[pbio.2000123] 
```

Both commands should generate zip files in the export directory.  Unzip it and you should be able to see a series of .csv, .html, and other assets in the directory.

#### Notes

The project currently needs a lot of help since it's now on an unsupported Ruby version.  On modern OS X machines, for example, it has a lot of difficulties getting started on a fresh Ruby 2.3.6 build due to OpenSSL, ZIP, and other lesser capability issues.  If you're reviewing this and run into similar problems that I have, please feel free to reach out to me for workarounds.

#### Major UI changes

A lot of files that were CSV are now HTML.  It's not the prettiest, but I think it's better than what it looked like before.

---

#### Code Review Tasks:

**Author tasks** (delete tasks that don't apply to your PR, this list should be finished before code review):

- [] If I made any UI changes at all, I've let the entire QA team know in the JIRA ticket and in the Aperta QA hipchat room.
- [ ] I have set the correct component(s) in the JIRA ticket
- [ ] I have set an appropriate resolution in the JIRA ticket
- [ ] If I changed the database schema, I enforced database constraints.
- [ ] If I created a migration, I updated the base data.yml seeds file. [instructions](https://developer.plos.org/confluence/display/TAHI/Seeds+maintenance)
- [ ] If I modified `app/services/journal_factory.rb`, I updated the base data.yml seeds file. [instructions](https://developer.plos.org/confluence/display/TAHI/Seeds+maintenance)
- [ ] I have ensured that the Heroku Review App has successfully deployed and is ready for PO UAT.
- [ ] I added an entry to `## [Unreleased]` in CHANGELOG.md

If I modified any environment variables:
- [ ] I made a pull request to change the files on the [molten repo](https://github.com/PLOS/molten/tree/dev/pillar/aperta) {PR LINK}
- [ ] I double-checked the `app.json` file to make sure that the heroku review apps are still inheriting the correct environment variables from staging

If I need to migrate existing data:
- [ ] If a data-migration rake task is needed, the task is found in `lib/tasks/data_migrations` within the `data:migrate` namespace. Example task name: `aperta_9999_migration_description`
- [ ] If there are steps to take outside of `rake db:migrate` for Heroku or other environments, I added copy-pastable instructions to [the confluence release page](https://developer.plos.org/confluence/display/TAHI/Deployment+information+for+Release)
- [ ] I verified the data-migration's results with `rake db:test_migrations` (complicated migrations should also have real specs)
- [ ] I've talked through the ramifications of the data-migration with Product Owners in regards to deployment timing
- [ ] If I created a data migration, I added pre- and post-migration assertions.

**Reviewer tasks** (these should be checked or somehow noted before passing on to PO):
- [ ] I read through the JIRA ticket's AC before doing the rest of the review
- [ ] I ran the code (in the review environment or locally). I agree the running code fulfills the Acceptance Criteria as stated on the JIRA ticket
- [ ] I read the code; it looks good
- [ ] I have found the tests to be sufficient for both positive and negative test cases

### CODE REVIEW

File: app/views/export/generic.html.erb [hunk 1]
```diff
@@ -0,0 +1,28 @@
+<!DOCTYPE html>
+<html>
+<head>
+  <style>
+    .divider:nth-child(odd) {
```
Comment (noxmwalsh) [inline app/views/export/generic.html.erb:5]: This is mainly to show unique sets of data within a card.  It's not the prettiest thing, but it easier to read than without it

File: app/views/export/generic.html.erb [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+      background-color: #e3e3e3;
+    }
+  </style>
+  <title><%= "test" %></title>
+</head>
+<body>
+<% id_count = nil %>
+<% @data.each do |data| %>
+  <% if id_count == nil %>
+    <% id_count = data[0].to_i.truncate %>
```
Comment (noxmwalsh) [inline app/views/export/generic.html.erb:17]: Some of the ids are like '5.2' as well as 5, so we have to truncate it.

As a side note: thinking it through, I probably should of called id_count into something like 'previous_major_id' or something like that to make it clearer.

File: app/views/export/generic.html.erb [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    <div class="divider">
+  <% end %>
+  <% if id_count != data[0].to_i.truncate %>
+    </div><div class="divider">
+    <% id_count = data[0].to_i.truncate %>
+  <% end %>
+  <p><%= data[0] %> - <%= raw data[1] %>: <span style="color:#137300"><%= data[2] %></span></p>
```
Comment (noxmwalsh) [inline app/views/export/generic.html.erb]: One aspect that made the old CSV harder to read was that a bunch of "answers" had HTML within them, so now we can render them fully.  This has the advantage of allowing href links to be clickable (although some of them might point to the app itself, which obviously isn't available), allow basic styling, and all the other goodness that HTML rendering provides.

File: lib/tasks/export.rake [hunk 4]
```diff
@@ -21,6 +21,7 @@
 require 'zip'
 require 'fileutils'
 require 'eml_to_pdf'
+require 'openssl'
```
Comment (benjaminkreen) [inline lib/tasks/export.rake]: whats up with this?

Comment (benjaminkreen) [reply lib/tasks/export.rake]: ah, just saw your comment, nvm

File: app/views/export/generic.html.erb (hunk 3 above)
Comment (benjaminkreen) [reply app/views/export/generic.html.erb]: you might want to `raw data[2]` as well? I've seem some escaped html here

File: lib/tasks/export.rake [hunk 5]
```diff
@@ -219,18 +237,17 @@ def export_paper(paper)
...
+          zos << export_question(data, snapshot.contents)
         end
       end
     end
   end
 end
 
 namespace :export do
-  task :manuscript_zip, [:short_doi] => [:environment] do |_, args|
+  task :manuscript_zip, [:short_doi] => [:environment] do |_, args  |
```
Comment (benjaminkreen) [inline lib/tasks/export.rake]: random space

Comment (noxmwalsh) [reply lib/tasks/export.rake]: Will correct

File: app/views/export/generic.html.erb (hunk 3 above)
Comment (noxmwalsh) [reply app/views/export/generic.html.erb]: Yeah, I don't see any harm in doing that there too.  There's likely certain answers that likely have HTML in it.

**Review by benjaminkreen — APPROVED** [review]
> For all cards, any response field is meaningfully labeled with context similar to what the user saw when filling out the card

Looks good and works, but the AC is hella vague, but as long as shane is cool with it then :shipit:

**Pull request merged**